compiler/modules/nback/function.bal

```diff
@@ -41,83 +41,128 @@ final RuntimeFunction addToRestArgsFunction = {
+        return { ref: operand.value, func: scaffold.getFunctionDefn(funcSymbol.identifier) };
+    }
+    return { ref: operand.value, func: check buildFunctionDecl(scaffold, funcSymbol, erasedSignature) };
 }
 
-function finishBuildCall(llvm:Builder builder, Scaffold scaffold, llvm:Function|llvm:PointerValue func,
-                         t:SemType[] paramTypes, t:SemType[] instantiatedParamTypes, bir:Operand[] args,
-                         t:SemType returnTy, bir:Register result) returns BuildError? {
-    llvm:Value[] argValues = check buildFunctionCallArgs(builder, scaffold, paramTypes, instantiatedParamTypes, args);
-    llvm:Value? retValue = buildFunctionCall(builder, scaffold, func, argValues);
-    buildStoreRet(builder, scaffold, semTypeRetRepr(returnTy), retValue, result);
+function functionValuePtrType(Scaffold scaffold, t:SemType funcType) returns llvm:PointerType {
+    t:FunctionAtomicType? atomic = t:functionAtomicType(scaffold.typeContext(), funcType);
+    if atomic == () {
+        // This is an approximation, which is sufficient for making the uniform call
+        return llvm:pointerType(llFunctionType);
+    }
+    t:FunctionSignature signature = t:functionSignature(scaffold.typeContext(), atomic);
+    return llvm:pointerType(functionValueType(signature));
 }
 
-// This is using the calling scheme defined in https://github.com/ballerina-platform/nballerina/issues/907#issuecomment-1041053503 (for inexact calls).
-// This function converts direct representation of call site type to uniform representation and call the uniformFunction
-// `createUniformFunction` defined in `init.bal` converts the uniform representation to direct representation of function definition,
-// call the function pointer and return the result in uniform representation.
-// Then this function convert that result to direct representation of call site type.
-function finishBuildCallIndirectInexact(llvm:Builder builder, Scaffold scaffold, bir:CallIndirectInsn insn,
-                                        llvm:BasicBlock afterCall, llvm:PointerValue funcValuePtr,
-                                        t:FunctionSignature signature) returns BuildError? {
-    var { returnType, paramTypes, restParamType } = signature;
-    int requiredArgCount = restParamType == () ? paramTypes.length() : paramTypes.length() - 1;
-    llvm:Value[] uniformArgs = from int i in 1 ..< requiredArgCount + 1
+function buildDirectCall(llvm:Builder builder, Scaffold scaffold, llvm:Function|llvm:PointerValue func,
+                         bir:CallInsn insn, t:SemType returnTy) returns BuildError? {
+    llvm:Value[] args = check buildDirectCallArgs(builder, scaffold, insn);
+    llvm:Value? retValue = buildFunctionCall(builder, scaffold, func, args);
+    buildStoreRet(builder, scaffold, semTypeRetRepr(returnTy), retValue, insn.result);
+}
+
+function buildDirectCallArgs(llvm:Builder builder, Scaffold scaffold, bir:CallInsn insn) returns llvm:Value[]|BuildError {
```

Comment:
This should take the signature and the args instead of the insn.

Reply:
Fixed